File: typescript/packages/core/tests/agent.test.ts
/**
 * Unit tests for the `Agent` component (spore-core issue #2).
 *
 * Mirrors `rust/crates/spore-core/src/agent.rs#tests` — same rules, same
 * verdicts, parallel structure.
 */

import { describe, expect, it } from "vitest";
import {
  AgentId,
  AgentModelError,
  EmptyResponse,
  MalformedToolCall,
  MockModelInterface,
  ModelAgent,
  ReplayModelInterface,
  Timeout,
  type Context,
  type ContentBlock,
  type ModelResponse,
  type ProviderInfo,
  type StopReason,
  type TokenUsage,
  type ToolCall,
} from "../src/index.js";

const provider: ProviderInfo = {
  name: "test",
  model_id: "test-1",
  context_window: 1000,
};

function usage(inT: number, outT: number): TokenUsage {
  return {
    input_tokens: inT,
    output_tokens: outT,
    cache_read_tokens: null,
    cache_write_tokens: null,
  };
}

function ctxUser(text: string): Context {
  return {
    messages: [{ role: "user", content: { type: "text", text } }],
    tools: [],
    params: { stop_sequences: [] },
  };
}

function textResp(text: string, stop: StopReason = "end_turn"): ModelResponse {
  return {
    content: [{ type: "text", text }],
    usage: usage(3, 4),
    stop_reason: stop,
  };
}

function toolResp(calls: ToolCall[]): ModelResponse {
  const content: ContentBlock[] = calls.map((c) => ({
    type: "tool_use" as const,
    id: c.id,
    name: c.name,
    input: c.input,
  }));
  return {
    content,
    usage: usage(5, 6),
    stop_reason: "tool_use",
  };
}

function makeAgent(): { agent: ModelAgent; model: MockModelInterface } {
  const model = new MockModelInterface(provider);
  const agent = new ModelAgent(AgentId.of("coding-agent"), model);
  return { agent, model };
}

describe("Agent — single-turn classification", () => {
  it("rule: one turn = exactly one model call", async () => {
    const { agent, model } = makeAgent();
    model.pushResponse(textResp("ok"));
    await agent.turn(ctxUser("hi"));
    expect(model.callCount).toBe(1);
  });

  it("classifies stop_reason=end_turn with text as FinalResponse", async () => {
    const { agent, model } = makeAgent();
    model.pushResponse(textResp("hello world"));
    const result = await agent.turn(ctxUser("hi"));
    expect(result.kind).toBe("final_response");
    if (result.kind === "final_response") {
      expect(result.content).toBe("hello world");
      expect(result.usage.input_tokens).toBe(3);
      expect(result.usage.output_tokens).toBe(4);
    }
  });

  it("classifies stop_reason=tool_use as ToolCallRequested", async () => {
    const { agent, model } = makeAgent();
    model.pushResponse(toolResp([{ id: "call_1", name: "read_file", input: { path: "/x" } }]));
    const result = await agent.turn(ctxUser("read /x"));
    expect(result.kind).toBe("tool_call_requested");
    if (result.kind === "tool_call_requested") {
      expect(result.calls).toHaveLength(1);
      expect(result.calls[0]!.id).toBe("call_1");
      expect(result.calls[0]!.name).toBe("read_file");
      expect(result.usage.input_tokens).toBe(5);
    }
  });

  it("ToolCallRequested may carry multiple parallel tool calls", async () => {
    const { agent, model } = makeAgent();
    model.pushResponse(
      toolResp([
        { id: "a", name: "read_file", input: { path: "/a" } },
        { id: "b", name: "read_file", input: { path: "/b" } },
      ]),
    );
    const result = await agent.turn(ctxUser("read both"));
    expect(result.kind).toBe("tool_call_requested");
    if (result.kind === "tool_call_requested") {
      expect(result.calls).toHaveLength(2);
      expect(result.calls[0]!.id).toBe("a");
      expect(result.calls[1]!.id).toBe("b");
    }
  });

  it("returns EmptyResponse when model returns no content blocks", async () => {
    const { agent, model } = makeAgent();
    model.pushResponse({
      content: [],
      usage: usage(1, 0),
      stop_reason: "end_turn",
    });
    const result = await agent.turn(ctxUser("?"));
    expect(result.kind).toBe("error");
    if (result.kind === "error") {
      expect(result.error).toBeInstanceOf(EmptyResponse);
      expect(result.error.kind).toBe("empty_response");
      expect(result.usage?.input_tokens).toBe(1);
    }
  });

  it("discards thinking blocks — thinking-only stays EmptyResponse", async () => {
    const { agent, model } = makeAgent();
    model.pushResponse({
      content: [{ type: "thinking", text: "musing" }],
      usage: usage(1, 2),
      stop_reason: "end_turn",
    });
    const result = await agent.turn(ctxUser("?"));
    expect(result.kind).toBe("error");
    if (result.kind === "error") {
      expect(result.error.kind).toBe("empty_response");
    }
  });

  it("surfaces ModelError wrapped in AgentError.ModelError with usage=null", async () => {
    const { agent, model } = makeAgent();
    model.pushError(new Timeout());
    const result = await agent.turn(ctxUser("hi"));
    expect(result.kind).toBe("error");
    if (result.kind === "error") {
      expect(result.error).toBeInstanceOf(AgentModelError);
      expect(result.error.kind).toBe("model_error");
      expect(result.usage).toBeNull();
      if (result.error instanceof AgentModelError) {
        expect(result.error.error.kind).toBe("timeout");
      }
    }
  });

  it("malformed when stop_reason=tool_use but no tool_use blocks", async () => {
    const { agent, model } = makeAgent();
    model.pushResponse({
      content: [{ type: "text", text: "hmm" }],
      usage: usage(2, 2),
      stop_reason: "tool_use",
    });
    const result = await agent.turn(ctxUser("?"));
    expect(result.kind).toBe("error");
    if (result.kind === "error") {
      expect(result.error).toBeInstanceOf(MalformedToolCall);
      expect(result.error.kind).toBe("malformed_tool_call");
      expect(result.usage).not.toBeNull();
    }
  });

  it("dispatches tool calls even when stop_reason=end_turn", async () => {
    const { agent, model } = makeAgent();
    model.pushResponse({
      content: [{ type: "tool_use", id: "x", name: "noop", input: {} }],
      usage: usage(1, 1),
      stop_reason: "end_turn",
    });
    const result = await agent.turn(ctxUser("?"));
    expect(result.kind).toBe("tool_call_requested");
  });

  it("classifies stop_reason=max_tokens with text as FinalResponse", async () => {
    const { agent, model } = makeAgent();
    model.pushResponse(textResp("truncated", "max_tokens"));
    const result = await agent.turn(ctxUser("?"));
    expect(result.kind).toBe("final_response");
  });

  it("classifies stop_reason=stop_sequence with text as FinalResponse", async () => {
    const { agent, model } = makeAgent();
    model.pushResponse(textResp("done.", "stop_sequence"));
    const result = await agent.turn(ctxUser("?"));
    expect(result.kind).toBe("final_response");
  });

  it("concatenates multiple text blocks", async () => {
    const { agent, model } = makeAgent();
    model.pushResponse({
      content: [
        { type: "text", text: "foo" },
        { type: "text", text: "bar" },
      ],
      usage: usage(1, 1),
      stop_reason: "end_turn",
    });
    const result = await agent.turn(ctxUser("?"));
    expect(result.kind).toBe("final_response");
    if (result.kind === "final_response") {
      expect(result.content).toBe("foobar");
    }
  });

  it("reports agent identity for tracing", () => {
    const model = new MockModelInterface(provider);
    const agent = new ModelAgent(AgentId.of("initializer"), model);
    expect(agent.id().asString()).toBe("initializer");
    expect(agent.id().equals(AgentId.of("initializer"))).toBe(true);
  });
});

describe("Agent — error JSON wire shape", () => {
  it("AgentError variants serialise with snake_case kind", () => {
    expect(new EmptyResponse().toJSON()).toEqual({ kind: "empty_response" });
    expect(new MalformedToolCall("x", "y").toJSON()).toEqual({
      kind: "malformed_tool_call",
      tool_name: "x",
      reason: "y",
    });
    const wrapped = new AgentModelError(new Timeout()).toJSON();
    expect(wrapped.kind).toBe("model_error");
    expect((wrapped.error as { kind: string }).kind).toBe("timeout");
  });
});

describe("Agent — replay against inline fixture", () => {
  it("classifies recorded exchanges consistently", async () => {
    const jsonl =
      `{"request":{"messages":[{"role":"user","content":{"type":"text","text":"call tool"}}],"tools":[],"params":{},"stream":false},"response":{"content":[{"type":"tool_use","id":"c1","name":"echo","input":{"v":1}}],"usage":{"input_tokens":3,"output_tokens":2},"stop_reason":"tool_use"},"provider":"anthropic"}\n` +
      `{"request":{"messages":[{"role":"user","content":{"type":"text","text":"finish"}}],"tools":[],"params":{},"stream":false},"response":{"content":[{"type":"text","text":"all done"}],"usage":{"input_tokens":4,"output_tokens":2},"stop_reason":"end_turn"},"provider":"anthropic"}`;
    const replay = ReplayModelInterface.fromJsonl(jsonl, {
      name: "anthropic",
      model_id: "replay",
      context_window: 200_000,
    });
    const agent = new ModelAgent(AgentId.of("replay-agent"), replay);

    const r1 = await agent.turn(ctxUser("call tool"));
    expect(r1.kind).toBe("tool_call_requested");
    if (r1.kind === "tool_call_requested") {
      expect(r1.calls).toHaveLength(1);
      expect(r1.calls[0]!.name).toBe("echo");
    }

    const r2 = await agent.turn(ctxUser("finish"));
    expect(r2.kind).toBe("final_response");
    if (r2.kind === "final_response") {
      expect(r2.content).toBe("all done");
    }
  });
});

describe("MockAgent", () => {
  it("returns queued results in FIFO order and counts calls", async () => {
    const { MockAgent } = await import("../src/index.js");
    const agent = new MockAgent(AgentId.of("mock"));
    agent.push({
      kind: "final_response",
      content: "one",
      usage: usage(1, 1),
    });
    const r = await agent.turn(ctxUser("?"));
    expect(r.kind).toBe("final_response");
    expect(agent.callCount).toBe(1);
    // Falls back to EmptyResponse when queue is empty.
    const r2 = await agent.turn(ctxUser("?"));
    expect(r2.kind).toBe("error");
    if (r2.kind === "error") {
      expect(r2.error.kind).toBe("empty_response");
    }
  });
});

File: go/spore-core/agent_test.go
package sporecore

import (
	"context"
	"encoding/json"
	"errors"
	"testing"
)

func agentProvider() ProviderInfo {
	return ProviderInfo{Name: "test", ModelID: "test-1", ContextWindow: 1000}
}

func userCtx(text string) Context {
	return Context{
		Messages: []Message{{Role: RoleUser, Content: NewTextContent(text)}},
		Tools:    []ToolSchema{},
		Params:   ModelParams{},
	}
}

func usage(in, out uint32) TokenUsage {
	return TokenUsage{InputTokens: in, OutputTokens: out}
}

func textRespEndTurn(text string) ModelResponse {
	return ModelResponse{
		Content:    []ContentBlock{NewTextBlock(text)},
		Usage:      usage(3, 4),
		StopReason: StopEndTurn,
	}
}

func toolResp(calls []ToolCall) ModelResponse {
	blocks := make([]ContentBlock, 0, len(calls))
	for _, c := range calls {
		blocks = append(blocks, NewToolUseBlock(c))
	}
	return ModelResponse{
		Content:    blocks,
		Usage:      usage(5, 6),
		StopReason: StopToolUse,
	}
}

func makeAgent(m *MockModel) *ModelAgent {
	return NewModelAgent(AgentID("coding-agent"), m)
}

// Rule: one turn = one model call.
func TestTurnMakesExactlyOneModelCall(t *testing.T) {
	m := NewMockModel(agentProvider())
	m.PushResponse(textRespEndTurn("ok"))
	agent := makeAgent(m)
	_ = agent.Turn(context.Background(), userCtx("hi"))
	if m.CallCount() != 1 {
		t.Fatalf("call count = %d, want 1", m.CallCount())
	}
}

// Rule: FinalResponse classification on stop_reason=EndTurn with text.
func TestFinalResponseOnEndTurnWithText(t *testing.T) {
	m := NewMockModel(agentProvider())
	m.PushResponse(textRespEndTurn("hello world"))
	agent := makeAgent(m)
	r := agent.Turn(context.Background(), userCtx("hi"))
	if r.Kind != TurnFinalResponse {
		t.Fatalf("kind = %q, want %q", r.Kind, TurnFinalResponse)
	}
	if r.Content != "hello world" {
		t.Fatalf("content = %q", r.Content)
	}
	if r.Usage == nil || r.Usage.InputTokens != 3 || r.Usage.OutputTokens != 4 {
		t.Fatalf("usage = %+v", r.Usage)
	}
}

// Rule: ToolCallRequested classification on stop_reason=ToolUse.
func TestToolCallRequestedOnToolUseStop(t *testing.T) {
	m := NewMockModel(agentProvider())
	m.PushResponse(toolResp([]ToolCall{
		{ID: "call_1", Name: "read_file", Input: json.RawMessage(`{"path":"/x"}`)},
	}))
	agent := makeAgent(m)
	r := agent.Turn(context.Background(), userCtx("read /x"))
	if r.Kind != TurnToolCallRequested {
		t.Fatalf("kind = %q", r.Kind)
	}
	if len(r.Calls) != 1 || r.Calls[0].ID != "call_1" || r.Calls[0].Name != "read_file" {
		t.Fatalf("calls = %+v", r.Calls)
	}
	if r.Usage == nil || r.Usage.InputTokens != 5 {
		t.Fatalf("usage = %+v", r.Usage)
	}
}

// Rule: ToolCallRequested may carry multiple parallel tool calls.
func TestToolCallRequestedCarriesMultipleCalls(t *testing.T) {
	m := NewMockModel(agentProvider())
	m.PushResponse(toolResp([]ToolCall{
		{ID: "a", Name: "read_file", Input: json.RawMessage(`{"path":"/a"}`)},
		{ID: "b", Name: "read_file", Input: json.RawMessage(`{"path":"/b"}`)},
	}))
	agent := makeAgent(m)
	r := agent.Turn(context.Background(), userCtx("both"))
	if r.Kind != TurnToolCallRequested || len(r.Calls) != 2 {
		t.Fatalf("got %+v", r)
	}
	if r.Calls[0].ID != "a" || r.Calls[1].ID != "b" {
		t.Fatalf("calls = %+v", r.Calls)
	}
}

// Rule: a clean EndTurn with no text and no tool calls is the model's
// voluntary completion signal → a (possibly empty) terminal FinalResponse,
// NOT an EmptyResponse error.
func TestCleanEndTurnWithNoContentIsEmptyFinalResponse(t *testing.T) {
	m := NewMockModel(agentProvider())
	m.PushResponse(ModelResponse{
		Content:    []ContentBlock{},
		Usage:      usage(1, 0),
		StopReason: StopEndTurn,
	})
	agent := makeAgent(m)
	r := agent.Turn(context.Background(), userCtx("?"))
	if r.Kind != TurnFinalResponse {
		t.Fatalf("expected empty FinalResponse, got %+v / err=%+v", r, r.Err)
	}
	if r.Content != "" {
		t.Fatalf("content = %q, want empty", r.Content)
	}
	if r.Reasoning != nil {
		t.Fatalf("reasoning = %+v, want nil", r.Reasoning)
	}
	if r.Usage.InputTokens != 1 {
		t.Fatalf("usage = %+v", r.Usage)
	}
}

// Rule: a truncated/abnormal empty (MaxTokens with no content) remains a
// genuine EmptyResponse error — only clean EndTurn is reclassified.
func TestMaxTokensWithNoContentIsEmptyResponse(t *testing.T) {
	m := NewMockModel(agentProvider())
	m.PushResponse(ModelResponse{
		Content:    []ContentBlock{},
		Usage:      usage(1, 0),
		StopReason: StopMaxTokens,
	})
	agent := makeAgent(m)
	r := agent.Turn(context.Background(), userCtx("?"))
	if r.Kind != TurnError || r.Err == nil || r.Err.Kind != AgentErrEmptyResponse {
		t.Fatalf("got %+v / err=%+v", r, r.Err)
	}
	if r.Usage == nil || r.Usage.InputTokens != 1 {
		t.Fatalf("usage = %+v", r.Usage)
	}
}

// Rule: Thinking-only content is treated as empty. Under MaxTokens (a
// truncated stop) this remains an EmptyResponse error; thinking is not a
// terminal response.
func TestThinkingBlocksDoNotSatisfyFinalResponse(t *testing.T) {
	m := NewMockModel(agentProvider())
	m.PushResponse(ModelResponse{
		Content:    []ContentBlock{NewThinkingBlock("musing")},
		Usage:      usage(1, 2),
		StopReason: StopMaxTokens,
	})
	agent := makeAgent(m)
	r := agent.Turn(context.Background(), userCtx("?"))
	if r.Kind != TurnError || r.Err.Kind != AgentErrEmptyResponse {
		t.Fatalf("got %+v", r)
	}
}

// Rule: ModelError surfaces wrapped, usage=nil.
func TestModelErrorSurfacesWrapped(t *testing.T) {
	m := NewMockModel(agentProvider())
	m.PushError(NewTimeout())
	agent := makeAgent(m)
	r := agent.Turn(context.Background(), userCtx("hi"))
	if r.Kind != TurnError {
		t.Fatalf("kind = %q", r.Kind)
	}
	if r.Err == nil || r.Err.Kind != AgentErrModelError || r.Err.ModelError == nil {
		t.Fatalf("err = %+v", r.Err)
	}
	if r.Err.ModelError.Kind != ModelErrTimeout {
		t.Fatalf("inner kind = %q", r.Err.ModelError.Kind)
	}
	if r.Usage != nil {
		t.Fatalf("usage should be nil for model error, got %+v", r.Usage)
	}
	// errors.As should unwrap through AgentError to *ModelError.
	var me *ModelError
	if !errors.As(r.Err, &me) {
		t.Fatalf("errors.As to *ModelError failed")
	}
}

// Rule: stop_reason=ToolUse without ToolUse blocks → MalformedToolCall.
func TestMalformedWhenToolUseStopButNoToolBlocks(t *testing.T) {
	m := NewMockModel(agentProvider())
	m.PushResponse(ModelResponse{
		Content:    []ContentBlock{NewTextBlock("hmm")},
		Usage:      usage(2, 2),
		StopReason: StopToolUse,
	})
	agent := makeAgent(m)
	r := agent.Turn(context.Background(), userCtx("?"))
	if r.Kind != TurnError || r.Err == nil || r.Err.Kind != AgentErrMalformedToolCall {
		t.Fatalf("got %+v", r)
	}
	if r.Usage == nil {
		t.Fatalf("expected usage on malformed tool call")
	}
}

// Tool calls present with non-ToolUse stop_reason are still dispatched.
func TestToolCallsDispatchedEvenWhenStopIsEndTurn(t *testing.T) {
	m := NewMockModel(agentProvider())
	m.PushResponse(ModelResponse{
		Content: []ContentBlock{
			NewToolUseBlock(ToolCall{ID: "x", Name: "noop", Input: json.RawMessage(`{}`)}),
		},
		Usage:      usage(1, 1),
		StopReason: StopEndTurn,
	})
	agent := makeAgent(m)
	r := agent.Turn(context.Background(), userCtx("?"))
	if r.Kind != TurnToolCallRequested || len(r.Calls) != 1 {
		t.Fatalf("got %+v", r)
	}
}

// Agent identity is reported.
func TestAgentIDReported(t *testing.T) {
	m := NewMockModel(agentProvider())
	agent := NewModelAgent(AgentID("initializer"), m)
	if agent.ID() != AgentID("initializer") {
		t.Fatalf("id = %q", agent.ID())
	}
}

// MaxTokens stop classifies as FinalResponse with text.
func TestMaxTokensStopClassifiesAsFinalResponse(t *testing.T) {
	m := NewMockModel(agentProvider())
	m.PushResponse(ModelResponse{
		Content:    []ContentBlock{NewTextBlock("truncated")},
		Usage:      usage(2, 5),
		StopReason: StopMaxTokens,
	})
	agent := makeAgent(m)
	r := agent.Turn(context.Background(), userCtx("?"))
	if r.Kind != TurnFinalResponse || r.Content != "truncated" {
		t.Fatalf("got %+v", r)
	}
}

func TestStopSequenceClassifiesAsFinalResponse(t *testing.T) {
	m := NewMockModel(agentProvider())
	m.PushResponse(ModelResponse{
		Content:    []ContentBlock{NewTextBlock("done.")},
		Usage:      usage(2, 1),
		StopReason: StopStopSequence,
	})
	agent := makeAgent(m)
	r := agent.Turn(context.Background(), userCtx("?"))
	if r.Kind != TurnFinalResponse {
		t.Fatalf("got %+v", r)
	}
}

// Multiple text blocks are concatenated.
func TestMultipleTextBlocksAreConcatenated(t *testing.T) {
	m := NewMockModel(agentProvider())
	m.PushResponse(ModelResponse{
		Content:    []ContentBlock{NewTextBlock("foo"), NewTextBlock("bar")},
		Usage:      usage(1, 1),
		StopReason: StopEndTurn,
	})
	agent := makeAgent(m)
	r := agent.Turn(context.Background(), userCtx("?"))
	if r.Kind != TurnFinalResponse || r.Content != "foobar" {
		t.Fatalf("got %+v", r)
	}
}

// JSON round-trip for cross-language fixture portability.
func TestTurnResultRoundtripJSON(t *testing.T) {
	cases := []TurnResult{
		NewToolCallRequested(
			[]ToolCall{{ID: "1", Name: "x", Input: json.RawMessage(`{"a":1}`)}},
			usage(2, 3),
		),
		NewFinalResponse("hi", usage(1, 1)),
		NewTurnError(NewEmptyResponseError(), nil),
		NewTurnError(NewMalformedToolCallError("foo", "bad"), &TokenUsage{InputTokens: 2}),
		NewTurnError(NewModelAgentError(NewTimeout()), nil),
	}
	for i, r := range cases {
		data, err := json.Marshal(r)
		if err != nil {
			t.Fatalf("case %d: marshal: %v", i, err)
		}
		var back TurnResult
		if err := json.Unmarshal(data, &back); err != nil {
			t.Fatalf("case %d: unmarshal %s: %v", i, data, err)
		}
		if back.Kind != r.Kind {
			t.Fatalf("case %d: kind mismatch %q vs %q", i, back.Kind, r.Kind)
		}
		// Re-marshal and compare bytes to detect representation drift.
		again, err := json.Marshal(back)
		if err != nil {
			t.Fatalf("case %d: re-marshal: %v", i, err)
		}
		if string(again) != string(data) {
			t.Fatalf("case %d: round-trip drift\n a=%s\n b=%s", i, data, again)
		}
	}
}

// JSON tag values match Rust serde_snake_case.
func TestTurnResultJSONTagsSnakeCase(t *testing.T) {
	r := NewFinalResponse("ok", usage(1, 1))
	data, err := json.Marshal(r)
	if err != nil {
		t.Fatal(err)
	}
	var probe map[string]interface{}
	if err := json.Unmarshal(data, &probe); err != nil {
		t.Fatal(err)
	}
	if probe["kind"] != "final_response" {
		t.Fatalf("kind tag = %v", probe["kind"])
	}
}

// AgentError variant tags match Rust.
func TestAgentErrorJSONShape(t *testing.T) {
	e := NewMalformedToolCallError("x", "y")
	data, err := json.Marshal(e)
	if err != nil {
		t.Fatal(err)
	}
	want := `{"kind":"MalformedToolCall","tool_name":"x","reason":"y"}`
	if string(data) != want {
		t.Fatalf("got %s", data)
	}

	// EmptyResponse
	e2 := NewEmptyResponseError()
	data2, _ := json.Marshal(e2)
	if string(data2) != `{"kind":"EmptyResponse"}` {
		t.Fatalf("empty: %s", data2)
	}

	// ModelError (transparent — inner kind takes over)
	e3 := NewModelAgentError(NewTimeout())
	data3, _ := json.Marshal(e3)
	if string(data3) != `{"kind":"Timeout"}` {
		t.Fatalf("model: %s", data3)
	}
}

// Context.IntoRequest preserves messages/tools/params and sets stream=false.
func TestContextIntoRequest(t *testing.T) {
	c := userCtx("hello")
	req := c.IntoRequest()
	if len(req.Messages) != 1 || req.Stream {
		t.Fatalf("req = %+v", req)
	}
}

// AgentError implements the error interface and is non-empty for every kind.
func TestAgentErrorString(t *testing.T) {
	variants := []*AgentError{
		NewModelAgentError(NewTimeout()),
		NewEmptyResponseError(),
		NewMalformedToolCallError("x", "y"),
	}
	for _, v := range variants {
		if v.Error() == "" {
			t.Fatalf("empty Error() for kind %q", v.Kind)
		}
	}
}

// Non-ModelError Go errors from the model are wrapped into ProviderError.
func TestNonModelErrorWrappedAsProviderError(t *testing.T) {
	m := NewMockModel(agentProvider())
	m.PushError(errors.New("boom"))
	agent := makeAgent(m)
	r := agent.Turn(context.Background(), userCtx("hi"))
	if r.Kind != TurnError || r.Err.Kind != AgentErrModelError {
		t.Fatalf("got %+v", r)
	}
	if r.Err.ModelError == nil || r.Err.ModelError.Kind != ModelErrProviderError {
		t.Fatalf("inner = %+v", r.Err.ModelError)
	}
	if r.Err.ModelError.Message != "boom" {
		t.Fatalf("msg = %q", r.Err.ModelError.Message)
	}
}
